### ISSUE TITLE
MNIST benchmark and accuracy question

### ISSUE THREAD

**Issue by dnat112** [mc2-project/delphi#24]
Would you be able to upload the pretrained values for the MNIST CNN that Delphi and MUSE are evaluated on in the MUSE paper? I see delphi/python/minionn/pretrained  (which I believe is the CIFAR-10 benchmark MUSE is evaluated on, correct?) and delphi/python/resnet/pretrained but not anything for the MNIST CNN.

Can you also confirm the accuracy that Delphi and MUSE achieved in the MUSE paper for the MNIST and CIFAR-10 networks?

Thank you!

**Comment by ryanleh** [comment]
Hi @dnat112, 

In Muse, unlike Delphi, accuracy wasn't really a concern since we weren't doing any approximations (besides fixed-point representation, but that is very accurate for the bit size we were using). As a result, while we did run correctness experiments, we didn't have any particular pre-trained network that we used for benchmarks: model weights were randomly generated.

Unless I'm forgetting, I don't believe we evaluated MNIST at all in Delphi? For CIFAR-10, Delphi's accuracy depends on how many approximations were being used -- could you clarify? In Muse, the accuracy would be the same as the underlying plaintext network you use (with maybe a small <1% error due to the fixed-point representation).

**Comment by dnat112** [comment]
Thanks for your response! To keep things less confusing, I'll use "Delphi" and "Muse" to refer to the solutions themselves, and "Delphi paper" and "Muse paper" to refer to the papers of the same name.

It appears that in Table 3 of the [Muse paper](https://www.usenix.org/system/files/sec21fall-lehmkuhl.pdf) both Delphi and Muse were evaluated on both MNIST and CIFAR-10. I am specifically asking about the accuracy for whatever configurations were used in the Muse paper for the MNIST and CIFAR-10 experiments (which I assume is with no approximate activation layers for Delphi, i.e., # of ReLUs replaced with squares = 0). 

If you don't have direct accuracy numbers for the Delphi and Muse MNIST and CIFAR-10 Muse paper experiment networks themselves, do you happen to know the accuracy of the baseline cleartext (presumably, non-fixed-point) networks you used? If I understand you correctly, I should be able to interpret 99% of this number as the lower-bound of the accuracy of Delphi and Muse for those experiments, correct? If so, how do you know that this error is only < 1% ? Is there a way you are able to bound the error?

Thank you!

**Comment by ryanleh** [comment]
Oh sorry, I understand you now 😄 

> I am specifically asking about the accuracy for whatever configurations were used in the Muse paper for the MNIST and CIFAR-10 experiments (which I assume is with no approximate activation layers for Delphi, i.e., # of ReLUs replaced with squares = 0).

We used random weights for the benchmarks in Table 3 of the Muse paper. And yes, the Delphi numbers in that table refer to Delphi without any polynomial approximations.

> do you happen to know the accuracy of the baseline cleartext (presumably, non-fixed-point) networks you used?

I believe the CIFAR-10 model had a plaintext accuracy of 84.80% and the MNIST model had a plaintext accuracy of 99.1%. 

> If I understand you correctly, I should be able to interpret 99% of this number as the lower-bound of the accuracy of Delphi and Muse for those experiments, correct? 

You can treat it as a lower bound, yes. To be a bit more specific, it's an approximate lower-bound on the number of times plaintext and secure inference differ from each-other on the testing set (which may improve or degrade accuracy).

> If so, how do you know that this error is only < 1% ? Is there a way you are able to bound the error?

We observed this lower bound experimentally: we took a pre-trained model, randomly chose a thousand inputs from the testing set (since it would take weeks to evaluate the entire testing set with secure inference haha), and checked that the plaintext and secure inference returned the same classification. For example, for the MiniONN CIFAR-10 network, secure inference gave the same classification as plaintext inference 99.88% of the time.

**Comment by dnat112** [comment]
Thanks! Just a few more questions on my end:

1. You say you had a pre-trained model for the MiniONN CIFAR-10 network -- is [this](https://github.com/mc2-project/delphi/tree/muse-experiments/python/minionn/pretrained/relu) the location of the model? If so, is there similar location for the MNIST model for the Muse MNIST experiment that you could point me to?

2. I see that the Muse paper replaced the max pooling in the MiniONN CIFAR-10 network with mean pooling. Was there a reason for this? 

3. 99.88%*84.80% = 84.7% is the "lower bound" I calculate for the accuracy for Delphi/MUSE over the CIFAR-10 dataset for the MUSE paper CIFAR-10 network structure. This is much higher than the 81.61% accuracy that MiniONN was able to achieve for the same dataset and same network structure (though possibly different weights). Do you have any idea where this difference is coming from? Is it from just better training of the initial model, or is it from some superior data representation strategy used by Delphi/Muse?

**Comment by ryanleh** [comment]
1. Yup! And no, I don't believe I have access to the MNIST model anymore since it was never pushed to a repo, sorry 😞. It should only take a few minutes to train a new one with comparable accuracy though.

2. Max pooling is a non-linear operation, mean pooling is a linear operation. The latter is much faster to compute in MPC 👍🏼 

3. Good observation, I believe MiniONN states that the plaintext accuracy they achieved was 81.61% vs. ours was 84.80%. So it would seem that we simply did a better job training the initial model.

**Comment by dnat112** [comment]
Thanks!

(Just to clarify, for MiniONN the 81.61% accuracy was for their final model. Not sure what the initial plaintext accuracy was that they started with (it wasn't reported in their paper as far as I could tell).)
